### ISSUE TITLE
How to access a record field in a loop

### ISSUE THREAD

**Issue by lpil** [michaeljones/gleam-templates#9]
Given this type

```rust
pub type Item {
  Item(text: String)
}
```
And this template
```rust
{> import themodule.{Item}
{> with items as List(Item)

{% for item in items %}
  {{ item.text }}
{% endfor %}
```
This error is returned when the code is compiled:

```
error: Unknown type for record access
   ┌─ ./src/todomvc/web/templates/home.gleam:62:50
   │
62 │     let builder = string_builder.append(builder, item.text)
   │                                                  ^^^^ I don't know what type this is
```
The problem here is that inside the loop the compiler can't (yet) know what the type is so there's no way to access values inside the record.

Is there a way to add type annotations to items in list being iterated over?

**Comment by michaeljones** [comment]
Nothing yet. I kind of hoped that providing top level types would be sufficient and everything else would be calculated from the compiler from there but it isn't a well tested assumption. 

It seems to generate:
```
import gleam/string_builder.{StringBuilder}
import gleam/list
import themodule.{Item}

pub fn render_builder(items items: List(Item), item item) -> StringBuilder {
  let builder = string_builder.from_string("")
  let builder = string_builder.append(builder, "
")
  let builder =
    list.fold(
      items,
      builder,
      fn(builder, item) {
        let builder = string_builder.append(builder, "
  ")
        let builder = string_builder.append(builder, item.text)
        let builder = string_builder.append(builder, "
")

        builder
      },
    )
  let builder = string_builder.append(builder, "
")

  builder
}

pub fn render(items items: List(Item), item item) -> String {
  string_builder.to_string(render_builder(items: items, item: item))
}
```
Where would we need more type annotations?

Edit: I'm guessing for the `fn` passed to `list.fold`? Sorry, I should check this out properly but I don't quite have time this evening. I can investigate properly this week.

**Comment by lpil** [comment]
Yes that's right. There's an issue to make this not required but it's not something we can do yet https://github.com/gleam-lang/gleam/issues/1093

**Comment by michaeljones** [comment]
Make sense. I think we could approach in two ways:

```rust
{> import themodule.{Item}
{> with items as List(Item)

{% for item in items %}
  {> with item as Item
  {{ item.text }}
{% endfor %}
```
(before or after the `for` not sure.)

or perhaps more neatly as:
```rust
{> import themodule.{Item}
{> with items as List(Item)

{% for item as Item in items %}
  {{ item.text }}
{% endfor %}
```
Not quite ideal as my brain leans towards the `as` keyword being used for renaming things in JS but it matches the `with .. as ..` syntax so it is at least internally consistent to some degree. I'll poke around trying to get this one working.

**Comment by michaeljones** [comment]
Do you think #12 covers it for you? I might be missing other subtleties, I'm not sure. 

Thanks for trying it and reporting the issue :)

**Comment by lpil** [comment]
Looks perfect! Thank you

**Comment by michaeljones** [comment]
Released in 0.6.0

**Comment by lpil** [comment]
Wonderful, thank you!
